git-pr-merge

```diff
@@ -65,40 +87,95 @@ prmerge::setup() {
   fi
 }
 
-prmerge::prepare() {
-  echo merging "${feature}" to "${master}"
-
-  # Fetch from all remotes so we can check if local branch is up-to-date
-  git remote update >/dev/null
-
-  if [[ "$(git rev-parse "${master}")" != "$(git rev-parse "${master}@{upstream}")" ]]; then
-    printf '%s is not up-to-date. You should run:\n' "${master}" >&2
-    printf '  git checkout %s\n' "${master}" >&2
-    printf '  git pull\n' >&2
-    return 1
-  fi
-
+common::prepare() {
   # cd to root of repo as master may not contain CWD, but switch back
   # when we're done.
   # shellcheck disable=SC2064
   # We want to evaluate $(pwd) now and not later.
   trap "cd \"$(pwd)\"" EXIT
   cd "$(git rev-parse --show-toplevel)"
 
+  # Fetch from all remotes so we can check if local branch is up-to-date
+  git remote update >/dev/null
+}
+
+prmerge::prepare() {
+  common::prepare
+
+  local do_pull=false
+  if [[ "$(git rev-parse "${master}")" != "$(git rev-parse "${master}@{upstream}")" ]]; then
+    printf '\n%s is not up-to-date. Press enter to update:' "${master}"
+    read -r
+    do_pull=true
+  fi
+
   git checkout -q "${master}"
+  "${do_pull}" && git pull
+}
+
+prupdate::prepare() {
+  common::prepare
+
+  # If the head of ${feature} is a merge commit from previously running
+  # git-pr-update, reset that commit and force push before generating the
+  # next merge commit. This seems to stop github removing approval on a PR.
+  read -ra feature_parents < <(git show --no-patch --format='%P' "${feature}" --)
+  if (( ${#feature_parents[@]} == 2 )); then
+    printf '\nHEAD is a merge commit. Resetting and force pushing. Press enter:'
+    read -r
+    git reset --hard @^2
+    git push --force
+  fi
 }
 
 prmerge::merge() {
+  echo merging "${feature}" to "${master}"
+
   #open 'https://gist.github.com/rxaviers/7360908'
   #open 'https://gitmoji.carloscuesta.me'
 
-  merge_message="$(prmerge::_merge_message "${feature}" "${master}")"
+  # If the feature branch has been updated with git-pr-update, then the head
+  # of the feature branch will be a merge commit that has the master branch
+  # as the first parent. In that case, just fast-forward merge instead of
+  # creating a merge commit.
+  read -ra feature_parents < <(git show --no-patch --format='%P' "${feature}" --)
+  master_hash=$(git rev-parse HEAD)
```

Comment:
can this now, that master is guaranteed, pulled be `git rev-parse ${master}`?

Reply:
done